### PR TITLE
23 check values in register

### LINKED ISSUE #23: Check values in register
check all inputs in frontend already and display an error message.

### PR DESCRIPTION
**Pull request by PurpurTentakel97**
blocked by #37 
close #23

### CODE REVIEW

File: src/routes/register/+page.svelte [hunk 1]
```diff
@@ -205,15 +140,15 @@
...
 				<HeadlinePage>Registrieren</HeadlinePage>
 				<Spacer --height="3rem" />
 				<MessageWrapper>
 					<ErrorMessage message={loggedInMessage} />
 					<ErrorMessage message={usernameMessage} />
 					<ErrorMessage message={emailMessage} />
 					<ErrorMessage message={passwordMessage} />
-					{#each errorMessage as message}
+					{#each errorMessages as message}
 						<ErrorMessage {message} />
```
Comment (mgerhold) [inline src/routes/register/+page.svelte:152]: Should you perform some escaping here? For example, if the user types HTML code in the e-mail field, the error message will take that code as-is and show it on the page. Or does Svelte do this automatically?

File: src/lookup/loginLookup.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+type LoginDict = {
+    [key: string]: string;
+};
+
+const lookup: LoginDict = {
+    "Username/email or password missing": "Name / E-Mail oder Passwort fehlt.",
+    "Unknown username or email": "Name / E-mail oder Passwort ungültig.",
+    "Invalid password": "Name / E-mail oder Passwort ungültig.",
```
Comment (mgerhold) [inline src/lookup/loginLookup.ts:8]: I think the spaces around the forward slashes are a bit weird (I don't think this is a common thing to do).

File: src/routes/register/validation.ts [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+    const hasUpperCase: boolean = /[A-Z]/.test(trimmed);
+    if (!hasUpperCase) {
+        return {
+            valid: false,
+            message: 'Das Passwort muss mindestens ein großen Buchstaben enthalten.'
+        }
+    }
+
+    const hasNumber: boolean = /[1-9]/.test(trimmed);
```
[the remark this answers is not in the document]
Comment (PurpurTentakel97) [reply src/routes/register/validation.ts:159]: that was ChatGPT and I was tired. 🙈

File: src/routes/register/+page.svelte (hunk 1 above)
Comment (PurpurTentakel97) [reply src/routes/register/+page.svelte:153]: I think inside svelte you need the html tag to perform that.
But I look that up.

Comment (PurpurTentakel97) [reply src/routes/register/+page.svelte:159]: I really dont know what to do right here.
Some say svelte escapes automatically with `{ }`. But when I use them the error message just says `[object]` or something.
Some say I have to escape that myself.
Some say it works only in `+page.svelte` files.  Some say it works in all `.svelte` files.
Some say it works eben in `.ts` files.

Im not sure what to do right now.

File: src/routes/register/validation.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+export async function onMailChangedAsync(mail: string, fetch: Function): Promise<string | undefined> {
+    const trimmed: string = mail.trim();
+
+    if (trimmed.length === 0) {
+        return 'Das Feld "E-Mail" ist leer.';
+    }
+
+    const valid: boolean = /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(trimmed);
```
Comment (mgerhold) [inline src/routes/register/validation.ts:42]: Please make sure that this regex is sufficient. If in doubt, please use an external package to validate the e-mail address.

File: src/routes/register/validation.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        return 'Die Passwörter stimmen nicht überein.';
+    }
+
+    if (trimmed.length < 8) {
+        return 'Das Password muss mindestens 8 Zeichen enthalten.';
+    }
+
+    const hasLowerCase: boolean = /[a-z]/.test(trimmed);
+    if (!hasLowerCase) {
+        return 'Das Passwort muss mindestens ein Kleinbuchstaben enthalten.';
```
Comment (mgerhold) [inline src/routes/register/validation.ts:87]:
```suggestion
        return 'Das Passwort muss mindestens einen Kleinbuchstaben enthalten.';
```

File: src/routes/register/validation.ts [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    const hasUpperCase: boolean = /[A-Z]/.test(trimmed);
+    if (!hasUpperCase) {
+        return 'Das Passwort muss mindestens ein Großbuchstaben enthalten.';
```
Comment (mgerhold) [inline src/routes/register/validation.ts:92]:
```suggestion
        return 'Das Passwort muss mindestens einen Großbuchstaben enthalten.';
```

File: src/routes/register/+page.svelte (hunk 1 above)
Comment (PurpurTentakel97) [reply src/routes/register/+page.svelte:159]: As far as I understand it.
This Line should already make sure that html gets not interpreted.
https://github.com/TechStreamConference/website-frontend/blob/23-check-values-in-register/src/elements/message/errorMessage.svelte#L7

I need to make sure in all components that this is like this.

File: src/routes/register/validation.ts (hunk 4 above)
Comment (PurpurTentakel97) [reply src/routes/register/validation.ts:42]: move this into an issue

**Review by mgerhold — APPROVED** [review]

**Pull request merged**